core/src/main/java/io/grpc/internal/AbstractClientStream2.java

```diff
@@ -0,0 +1,300 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.annotation.Nullable;
+
+/**
+ * The abstract base class for {@link ClientStream} implementations. Extending classes only need to
+ * implement {@link #transportState()} and {@link #abstractClientStreamSink()}. Must only be called
+ * from the sending application thread.
+ */
+public abstract class AbstractClientStream2 extends AbstractStream2
+    implements ClientStream, MessageFramer.Sink {
+
+  private static final Logger log = Logger.getLogger(AbstractClientStream2.class.getName());
+
+  /**
+   * A sink for outbound operations, separated from the stream simply to avoid name
+   * collisions/confusion. Only called from application thread.
+   */
+  protected interface Sink {
+    /**
+     * Sends an outbound frame to the remote end point.
+     *
+     * @param frame a buffer containing the chunk of data to be sent, or {@code null} if {@code
+     *     endOfStream} with no data to send
+     * @param endOfStream {@code true} if this is the last frame; {@code flush} is guaranteed to be
+     *     {@code true} if this is {@code true}
+     * @param flush {@code true} if more data may not be arriving soon
+     */
+    void writeFrame(@Nullable WritableBuffer frame, boolean endOfStream, boolean flush);
+
+    /**
+     * Requests up to the given number of messages from the call to be delivered. This should end up
+     * triggering {@link TransportState#requestMessagesFromDeframer(int)} on the transport thread.
+     */
+    void request(int numMessages);
+
+    /**
+     * Tears down the stream, typically in the event of a timeout. This method may be called
+     * multiple times and from any thread.
+     *
+     * <p>This is a clone of {@link ClientStream#cancel()}.
+     */
+    void cancel(Status status);
+  }
+
+  private final MessageFramer framer;
+  private boolean outboundClosed;
+  /**
+   * Whether cancel() has been called. This is not strictly necessary, but removes the delay between
+   * cancel() being called and isReady() beginning to return false, since cancel is commonly
+   * processed asynchronously.
+   */
+  private volatile boolean cancelled;
+
+  protected AbstractClientStream2(WritableBufferAllocator bufferAllocator) {
+    framer = new MessageFramer(this, bufferAllocator);
+  }
+
+  @Override
+  protected abstract TransportState transportState();
+
+  @Override
+  public void start(ClientStreamListener listener) {
+    transportState().setListener(listener);
+  }
+
+  /**
+   * Sink for transport to be called to perform outbound operations. Each stream must have its own
+   * unique sink.
+   */
+  protected abstract Sink abstractClientStreamSink();
```

Comment:
It _isn't_ application state. Application state is in the stream itself, not the sink.

Actually, the name should have been abstractClientStream2Sink (I failed to fix it when I did some renaming), as in AbstractClientStream2.Sink. I'd rather not elide the abstract from the name because that is less precise, still prone to conflict, and it isn't used enough that the verbosity is an issue.